contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py

```diff
@@ -196,60 +189,99 @@ def checkstyle(self, interpreter, sources):
 
       with self.context.new_workunit(name='pythonstyle',
                                      labels=[WorkUnitLabel.TOOL, WorkUnitLabel.LINT],
-                                     cmd=' '.join(checker.cmdline(args))) as workunit:
+                                     cmd=safe_shlex_join(checker.cmdline(args))) as workunit:
         return checker.run(args=args,
-                                    stdout=workunit.output('stdout'),
-                                    stderr=workunit.output('stderr'))
-
-  def _constraints_are_whitelisted(self, constraint_tuple):
-    """
-    Detect whether a tuple of compatibility constraints
-    matches constraints imposed by the merged list of the global
-    constraints from PythonSetup and a user-supplied whitelist.
-    """
-    if self._acceptable_interpreter_constraints == []:
-      # The user wants to lint everything.
-      return True
-    return all(version.parse(constraint) in self._acceptable_interpreter_constraints
-           for constraint in constraint_tuple)
+                           stdout=workunit.output('stdout'),
+                           stderr=workunit.output('stderr'),
+                           # Overriding PEX_PYTHON_PATH is necessary to satisfy the interpreter
+                           # constraint we added to guard against a pexrc.
+                           env={'PEX_PYTHON_PATH': interpreter.binary})
+
+  class CheckstyleSetupError(TaskError): pass
+
+  def _parse_requirement(self, filt):
+    try:
+      return requirements.Requirement(filt)
+    except requirements.InvalidRequirement as orig_error:
+      # For the case of '>3', add something that looks like a requirement name.
+      new_filt = 'XXX{}'.format(filt)
+      try:
+        return requirements.Requirement(new_filt)
+      except requirements.InvalidRequirement:
+        raise self.CheckstyleSetupError('Could not parse interpreter constraint {}: {}'
+                                        .format(filt, str(orig_error)),
+                                        orig_error)
+
+  def _interpreter_is_py3(self, interpreter):
+    return interpreter.identity.version[0] == 3
+
+  def _partition_targets_by_min_interpreter(self, tgts):
```

Comment:
This feels like code we might already have somewhere. Would be good to wait for a John review before landing.

Reply:
Will do, github's list of my open PRs isn't going to let me forget about it and this is isn't blocking anything time-sensitive so we can easily wait until then.